SDK support for Content Versioning

azrikahar opened the pull request:
## Scope

What's changed:

- added `directus_versions` schema
- added CRUD operations for Content Versions
- added utils for Content Versions (compare, save, and promote)
- Update every SDK examples in Content Versions API reference

## Potential Risks / Drawbacks

- Perhaps a tiny increase of the SDK's codebase

## Review Notes / Questions

- May require more attention in `rest/commands/utils/versions.ts` in terms of the typings used, as I'm not sure did I form them correctly

---

Fixes #19999

azrikahar committed:
add directus_versions schema

azrikahar committed:
update collection and revision schema

azrikahar committed:
add CRUD operations

azrikahar committed:
add utils commands for versions

azrikahar committed:
remove "usually" from Promote a Version

azrikahar committed:
change "rename" to "update"

We used to only be able to update the "name" field, hence the idea was to use "rename" to make it clearly. However now we can update "key" + "name", so using the generic "update" is more appropriate.

br41nslug commented on sdk/src/rest/commands/utils/versions.ts:
```diff
@@ -0,0 +1,68 @@
+ * @returns All fields with different values, along with the hash of the main item and the information whether the current
+version is outdated (main item has been updated since the creation of the current version)
+ */
+export const compareVersion =
```
I feel the naming here is too vague and possibly even misleading. lets use a more feature specific and descriptive name here for these SDK functions like `compareDraftVersion` or `compareRevisionVersion`.

azrikahar replied:
Makes sense, `compareContentVersion` sounds fitting with Content Versioning 👍 In that case, I'll proceed with updating all of them such as `readVersion`/`updateVersion` to `readContentVersion`/`updateContentVersion` respectively as well.

azrikahar replied:
resolved via https://github.com/directus/directus/pull/20000/commits/e71199bf9555eb164c374ddbecd433d81ad6de15 🙌

azrikahar committed:
add SDK examples

azrikahar committed:
add changeset

azrikahar committed:
run prettier on docs

azrikahar committed:
update function naming with Content as prefix

paescuj committed:
Delete changeset

We're going to add a dedicated one for all changes related to Content Versioning

br41nslug commented on sdk/src/rest/commands/utils/versions.ts:
```diff
@@ -0,0 +1,68 @@
+			main: Item;
+		},
+		Schema
+	> =>
+	() => ({
+		path: `/versions/${id}/compare`,
+		method: 'GET',
+	});
```
I think we may want to check for empty `id` here too so we don't get `/versions//compare`
```suggestion
	> =>
	() => {
		throwIfEmpty(key, 'Key cannot be empty');
		
	    return {
    		path: `/versions/${id}/compare`,
	    	method: 'GET',
	    }
	};
```

paescuj replied:
Great catch ❤️

br41nslug commented on sdk/src/rest/commands/utils/versions.ts:
```diff
@@ -0,0 +1,68 @@
+		id: DirectusVersion<Schema>['id'],
+		item: Partial<Item>
+	): RestCommand<Item, Schema> =>
+	() => ({
+		path: `/versions/${id}/save`,
+		method: 'POST',
+		body: JSON.stringify(item),
+	});
```
Same ID check for empty here would be good

br41nslug commented on sdk/src/rest/commands/utils/versions.ts:
```diff
@@ -0,0 +1,68 @@
+		mainHash: string,
+		fields?: (keyof UnpackList<Item>)[]
+	): RestCommand<string | number, Schema> =>
+	() => ({
+		path: `/versions/${id}/promote`,
+		method: 'POST',
+		body: JSON.stringify(fields ? { mainHash, fields } : { mainHash }),
+	});
```
Same ID check for empty here would be good

br41nslug reviewed (dismissed):
Couple of small tweaks but the types and docs are looking great 🚀

paescuj committed:
Update method names in reference doc

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

paescuj committed:
Check for empty ids

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

paescuj committed:
Format code blocks

paescuj committed:
Switch back to `branches`

paescuj committed:
Update item link

azrikahar committed:
add version query to readItem & readSingleton

paescuj committed:
Revert "Switch back to `branches`"

This reverts commit 604487bd475f05135229aefa50e754fdbd6d4148.

paescuj committed:
Final version (pun intended)

paescuj committed:
Fix typo

paescuj committed:
Remove chars that snuck in

paescuj committed:
Another typo

paescuj committed:
Add ref to version param

paescuj committed:
Add GQL snippets

paescuj committed:
Format snippets
